rpc/client.go

```diff
@@ -25,72 +24,91 @@ type UploadReportOptions struct {
 
 func UploadReport(ctx context.Context, opts UploadReportOptions) error {
 	if err := opts.convertAndUploadArtifacts(ctx); err != nil {
-		return errors.Wrap(err, "problem uploading tests for report")
+		return errors.Wrap(err, "uploading tests for report")
 	}
 	return errors.Wrap(uploadTests(ctx, gopb.NewCedarPerformanceMetricsClient(opts.ClientConn), opts.Report, opts.Report.Tests, opts.DryRun),
-		"problem uploading tests for report")
+		"uploading tests for report")
 }
 
 func (opts *UploadReportOptions) convertAndUploadArtifacts(ctx context.Context) error {
-	jobQueue := queue.NewLocalLimitedSize(runtime.NumCPU(), len(opts.Report.Tests)*2)
+	var wg sync.WaitGroup
+	catcher := grip.NewBasicCatcher()
+	testChan := make(chan poplar.Test)
+	wg.Add(1)
+	go opts.artifactProducer(ctx, testChan, catcher, &wg)
+
+	workers := 1
 	if !opts.SerializeUpload {
-		if err := jobQueue.Start(ctx); err != nil {
-			return errors.Wrap(err, "problem starting artifact upload queue")
-		}
-		defer jobQueue.Runner().Close(ctx)
+		workers = runtime.NumCPU()
 	}
-
-	queue := make([]poplar.Test, len(opts.Report.Tests))
-	for i, test := range opts.Report.Tests {
-		queue[i] = test
+	for i := 0; i < workers; i++ {
+		wg.Add(1)
+		go opts.artifactConsumer(ctx, testChan, catcher, &wg)
 	}
 
-	for len(queue) != 0 {
-		test := queue[0]
-		queue = queue[1:]
+	wg.Wait()
+	return catcher.Resolve()
+}
+
+func (opts *UploadReportOptions) artifactProducer(ctx context.Context, testChan chan poplar.Test, catcher grip.Catcher, wg *sync.WaitGroup) {
+	defer func() {
+		catcher.Add(recovery.HandlePanicWithError(recover(), nil, "artifact upload producer"))
+		close(testChan)
+		wg.Done()
```

Comment:
You shouldn't need to add the producer to the waitgroup if you close the test channel - adding just the consumers to the wait group is sufficient to ensure all the work is processed (or the context is cancelled) when the wait group returns from waiting.